packages/remix-react/data.ts

```diff
@@ -35,23 +36,143 @@ export async function fetchData(
   routeId: string,
   signal: AbortSignal,
   submission?: Submission
-): Promise<Response | Error> {
+): Promise<[Response | Error, any]> {
   url.searchParams.set("_data", routeId);
+  let response: Response;
+  let events: {
+    [key: string]: {
+      promise: Promise<any>;
+      resolve: (value?: any) => void;
+      resolved?: boolean;
+    };
+  } = {};
+  if (submission) {
+    response = await fetch(url.href, getActionInit(submission, signal));
+  } else {
+    let ogResponse = await fetch(url.href, {
+      credentials: "same-origin",
+      signal,
+    });
+
+    response = ogResponse;
+    if (
+      !isRedirectResponse(ogResponse) &&
+      ogResponse.body &&
+      ogResponse.headers.get("Content-Type")?.includes("text/remix-deferred")
+    ) {
+      response = await new Promise<Response>(async (resolve, reject) => {
+        let reader = ogResponse.body!.getReader();
```

Comment:
Can we have a utility method that is responsible for the decode step here. Something like:

```ts
// Maybe like this?
let { response, events } = await decodeRemixDeferred(response.body);
```